[test_ring_joint_attention.py:184] | Sanity tests nightly debug run | blackhole nightly debug run (Ubuntu 22.04) / blackhole-multi-card-fast-unit-tests (BH-Quietbox-2) / blackhole BH-Quietbox-2 sdpa nightly tests (QB2 only)

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] Ring joint SDPA program cache test fails with assert 3 == 1 on Blackhole 2x2 mesh (SD3.5 config)
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** Sanity tests nightly debug run
**Job:** blackhole nightly debug run (Ubuntu 22.04) / blackhole-multi-card-fast-unit-tests (BH-Quietbox-2) / blackhole BH-Quietbox-2 sdpa nightly tests (QB2 only)

### Regression timeline

- **Last passing commit:** N/A (not provided)
- **First failing commit:** `5b69c0a2f4e5` — 2026-06-27T00:25:40Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/28272758612/job/83788444252)
- **Most recent failing commit:** `9a71e73f6fbc` — 2026-06-28T00:23:58Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/28306273824/job/83867852253)

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/28306273824/job/83867852253
- https://github.com/tenstorrent/tt-metal/actions/runs/28272758612/job/83788444252

### Failing test path(s)

- `tests/nightly/blackhole/sdpa/test_ring_joint_attention.py::test_ring_joint_sdpa_program_cache`
  - `[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf16]`
  - `[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf8_b]`
  - `[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf4_b]`

### Error excerpt (terminal failure from logs)

**Run 1 (2026-06-28):**
```
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_attention.py::test_ring_joint_sdpa_program_cache[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf16] - assert 3 == 1
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_attention.py::test_ring_joint_sdpa_program_cache[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf8_b] - assert 3 == 1
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_attention.py::test_ring_joint_sdpa_program_cache[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf4_b] - assert 3 == 1
>       assert submesh.cache_entries_counter.total == 1
E       assert 3 == 1
E        +  where 3 = <tests.tests_common.cache_entries_counter.CacheEntriesCounter object at 0x74c473b68cd0>.total
E        +    where <tests.tests_common.cache_entries_counter.CacheEntriesCounter object at 0x74c473b68cd0> = MeshDevice(2x2 grid, 4 devices).cache_entries_counter
tests/nightly/blackhole/sdpa/test_ring_joint_attention.py:184: AssertionError
============= 3 failed, 84 passed, 92 skipped in 668.32s (0:11:08) =============
```

**Run 2 (2026-06-27):**
```
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_attention.py::test_ring_joint_sdpa_program_cache[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf16] - assert 3 == 1
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_attention.py::test_ring_joint_sdpa_program_cache[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf8_b] - assert 3 == 1
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_attention.py::test_ring_joint_sdpa_program_cache[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf4_b] - assert 3 == 1
>       assert submesh.cache_entries_counter.total == 1
E       assert 3 == 1
E        +  where 3 = <tests.tests_common.cache_entries_counter.CacheEntriesCounter object at 0x74c473b68cd0>.total
E        +    where <tests.tests_common.cache_entries_counter.CacheEntriesCounter object at 0x74c473b68cd0> = MeshDevice(2x2 grid, 4 devices).cache_entries_counter
tests/nightly/blackhole/sdpa/test_ring_joint_attention.py:184: AssertionError
```

### Reproduction steps

On a Blackhole 2x2 mesh device (QB2):

```bash
pytest tests/nightly/blackhole/sdpa/test_ring_joint_attention.py -k "test_ring_joint_sdpa_program_cache and 2rpx2up and line and no_trace and sd35" -v
```

### Notes

The test asserts that after running ring joint SDPA, only **1** program cache entry is created (program cache hit on re-run). Instead, **3** entries are found (`assert 3 == 1`), indicating that the ring joint attention kernel for the SD3.5 configuration on Blackhole 2x2 mesh is not properly reusing cached programs — each dtype variant (bf16, bf8_b, bf4_b) appears to be generating a distinct cache entry instead of hitting a shared cached program.

Only the `no_trace` + `sd35` variants fail (3 failing out of 3 parameterized dtype variants). This is consistent across both runs.

**Determinism check:**
- Run 1 (commit `9a71e73f6fbc`, 2026-06-28): `FAILED tests/nightly/blackhole/sdpa/test_ring_joint_attention.py::test_ring_joint_sdpa_program_cache[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf16] - assert 3 == 1` (and bf8_b, bf4_b)
- Run 2 (commit `5b69c0a2f4e5`, 2026-06-27): `FAILED tests/nightly/blackhole/sdpa/test_ring_joint_attention.py::test_ring_joint_sdpa_program_cache[blackhole-2rpx2up-mesh_device0-line-2-no_trace-sd35-bf16] - assert 3 == 1` (and bf8_b, bf4_b)
- Both runs: identical test names, identical assertion (`assert 3 == 1`), identical file+line (`test_ring_joint_attention.py:184: AssertionError`), identical count (3 failed).

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: Sanity tests nightly debug run`
`Auto-triage-job-name: blackhole nightly debug run (Ubuntu 22.04) / blackhole-multi-card-fast-unit-tests (BH-Quietbox-2) / blackhole BH-Quietbox-2 sdpa nightly tests (QB2 only)`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

<!-- PIPELINE-REPORT-START -->
## Pipeline report
Analysis report: https://github.com/tenstorrent/tt-metal/actions/runs/28281152041
Slack notification: https://tenstorrent.slack.com/archives/C0B1F8Z7GQM/p1782542364246639?thread_ts=1782541544.216509&cid=C0B1F8Z7GQM
<!-- PIPELINE-REPORT-END -->

--- a/tests/nightly/blackhole/sdpa/test_ring_joint_attention.py
+++ b/tests/nightly/blackhole/sdpa/test_ring_joint_attention.py
@@ -2,8 +2,6 @@
 
 # SPDX-License-Identifier: Apache-2.0
 
-import torch
-
 import ttnn
 from loguru import logger
 import pytest
@@ -90,7 +88,7 @@ def test_ring_joint_sdpa_dit_bh_qb_ge(
     ],
     ids=["sd35"],
 )
-@pytest.mark.parametrize("n_iters, trace_enabled", [(1, False)], ids=["no_trace"])
+@pytest.mark.parametrize("n_iters, trace_enabled", [(3, False)], ids=["no_trace"])
 @pytest.mark.parametrize("num_links", [2])
 @pytest.mark.parametrize(
     "device_params, all_gather_topology",
@@ -148,37 +146,31 @@ def test_ring_joint_sdpa_program_cache(
 
     skip_check = False
 
-    dummy_tensors = []
-    for i in range(3):
-        dummy_tensors.append(
-            ttnn.from_torch(
-                torch.rand((b, nh, seq_len, d)),
-                device=submesh,
-                layout=ttnn.TILE_LAYOUT,
-                dtype=dtype,
-                mesh_mapper=ttnn.ShardTensor2dMesh(submesh, mesh_shape=tuple(submesh.shape), dims=[None, None]),
-            )
-        )
-
-        run_ring_joint_sdpa(
-            submesh,
-            b,
-            nh,
-            seq_len,
-            seq_len,
-            joint_seq_len,
-            d,
-            q_chunk_size,
-            k_chunk_size,
-            dtype,
-            n_iters,
-            trace_enabled,
-            num_links,
-            rp_axis,
-            up_axis,
-            all_gather_topology,
-            skip_check,
-            pcc_threshold,
-        )
+    # Run the op n_iters (>1) times within a SINGLE run_ring_joint_sdpa invocation. The op must be
+    # exercised under one sub-device-manager lifetime: run_ring_joint_sdpa loads a sub-device manager
+    # on entry, which clears the program cache (mesh_device.cpp clear_program_cache on manager switch),
+    # so looping the whole helper would clear the cache between calls and defeat the reuse check.
+    # run_ring_joint_sdpa's internal loop runs the op n_iters times with distinct per-iter persistent
+    # buffers and global semaphores, so cache reuse is still validated against address variation.
+    run_ring_joint_sdpa(
+        submesh,
+        b,
+        nh,
+        seq_len,
+        seq_len,
+        joint_seq_len,
+        d,
+        q_chunk_size,
+        k_chunk_size,
+        dtype,
+        n_iters,
+        trace_enabled,
+        num_links,
+        rp_axis,
+        up_axis,
+        all_gather_topology,
+        skip_check,
+        pcc_threshold,
+    )
 
     assert submesh.cache_entries_counter.total == 1

--- a/ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/sparse_sdpa_msa_compute.cpp
+++ b/ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/sparse_sdpa_msa_compute.cpp
@@ -134,7 +134,6 @@ void kernel_main() {
                             /*subblock_h=*/qsb,
                             /*inner_dim=*/DHt,
                             /*matmul_stride=*/DHt,
-                            /*trigger_reduce=*/false,
                             /*skip_pack_configure=*/true);
                     }
                     // Publish the band to UNPACK while holding wr_ptr for in-place sub_exp.
@@ -192,7 +191,6 @@ void kernel_main() {
                             /*subblock_h=*/qsb,
                             /*inner_dim=*/Skt,
                             /*matmul_stride=*/KT_stride,
-                            /*trigger_reduce=*/false,
                             /*skip_pack_configure=*/true);
                     }
                     pack_to_unpack_sync();                // publish held out_cur packs before flash combine

